File: c/insertion_sort_list.c
// Sort a linked list using insertion sort.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */

#include <stdio.h>

#include "lib/list.h"

typedef struct ListNode* Position;
typedef struct ListNode* List;

void Swap(Position P1, Position P2)
{
  int Temp;

  Temp = P1 -> val;
  P1 -> val = P2 -> val;
  P2 -> val = Temp;
}

struct ListNode* insertionSortList(struct ListNode* head) {
  if (head == NULL)
    return NULL;

  Position P1, P2, PtrNode;

  P1 = head -> next;
  P2 = head;
  while (P1 != NULL)
  {
    if (P1 -> val < P2 -> val)
    {
      PtrNode = head;
      while (PtrNode -> val < P1 -> val && PtrNode != P1)
        PtrNode = PtrNode -> next;

      if (PtrNode != P1)
      {
        P2 -> next = P1 -> next;
        Swap(P1, PtrNode);
        P1 -> next = PtrNode -> next;
        PtrNode -> next = P1;
      }
    }

    P2 = P1;
    P1 = P1 -> next;
  }

  return head;
}

int main(int argc, char const *argv[])
{
  List L;
  
  L = CreateRandList(16);
  PrintList(L);

  L = insertionSortList(L);
  PrintList(L);
  DestroyList(L);

  L = CreateReversedList(16);
  PrintList(L);

  L = insertionSortList(L);
  PrintList(L);
  DestroyList(L);

  L = CreateList(16);
  PrintList(L);

  L = insertionSortList(L);
  PrintList(L);
  DestroyList(L);

  return 0;
}

File: ruby/insertion_sort_list.rb
# Sort a linked list using insertion sort.

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {ListNode}

require_relative 'src/list.rb'

def insertion_sort_list(head)
	return head if head.nil? || head.next.nil?

	pre_node = head
	node = head.next

	until node.nil?
		if pre_node.val <= node.val
			pre_node = node
			node = node.next
			next
		end

		pre_node.next = node.next
		position = head
		position = position.next while position.val < node.val && position != node
		position.val, node.val = node.val, position.val
		node.next = position.next
		position.next = node
		node = pre_node.next
	end

	head
end

list = create_rand_list 2
print_list list
list = insertion_sort_list list
print_list list